internal/handlers/watch_cache.go

```diff
@@ -0,0 +1,175 @@
+	if cache == nil {
+		cache = &watchCache{
+			manager:    manager,
+			ringBuffer: make([]watchItem, cacheSize),
+		}
+		manager.tailCaches[key] = cache
+	}
+
+	cache.openCounter += 1
+	return &watchWorker{
+		cache:              cache,
+		fetchFn:            fetchFn,
+		fetchFromTailCache: false,
+	}
+}
+
+func (sub *watchWorker) Close() {
+	if sub.cache == nil {
+		return
+	}
+	sub.cache.manager.mu.Lock()
+	defer sub.cache.manager.mu.Unlock()
+	sub.cache.openCounter -= 1
+	sub.cache = nil
+}
+
+func (sub *watchWorker) Fetch(db *gorm.DB, gtRevision uint64) (WatchableList, error) {
+	if !sub.fetchFromTailCache {
+
+		wl, err := sub.fetchFn(db, gtRevision)
+		if err != nil {
+			return wl, err
+		}
+
+		// empty fetch means we have hit the tail... next fetch will load from the shared tail cache
+		if wl.Len() == 0 {
+			sub.fetchFromTailCache = true
+		}
+
+		return wl, nil
+
+	} else {
+		return sub.fetchFromSharedTailCache(db, gtRevision)
+	}
+}
+
+func (sub *watchWorker) fetchFromSharedTailCache(db *gorm.DB, gtRevision uint64) (WatchableList, error) {
+	for {
+		items, writePos := sub.fetchFromRing(gtRevision)
+		if len(items) != 0 {
+			return items, nil
+		}
+
+		if !sub.fetchFromTailCache {
+			// we can't read from ring buffer anymore, go back to using the Fetch method.
+			return sub.Fetch(db, gtRevision)
+		}
+
+		result, err := sub.tryFillRing(db, gtRevision, writePos)
+		if result != nil || err != nil {
+			return result, err
+		}
+		// we were not the sub that did the DB work, loop around
+		// to see if we can use the results of sub that did do the work.
+	}
+}
+
+func (sub *watchWorker) fetchFromRing(lastRevision uint64) (watchItemList, uint64) {
+	items := watchItemList{}
+
+	// get items of the current cache...
+	cache := sub.cache
+	cache.mu.RLock()
+	defer cache.mu.RUnlock()
+
+	writePos := cache.writePos
+	ringSize := uint64(len(cache.ringBuffer))
+	for sub.readPos != writePos {
+		i := sub.readPos % ringSize
+		item := cache.ringBuffer[i]
+		sub.readPos += 1
+		if lastRevision < item.Revision {
+			if lastRevision+1 == item.Revision {
+				items = append(items, item)
+				lastRevision = item.Revision
+			} else {
+				//fmt.Println("sub.fetchFromTailCache = false")
+				sub.fetchFromTailCache = false
+				break
+			}
+		}
+	}
+	return items, writePos
+}
+
+func (sub *watchWorker) tryFillRing(db *gorm.DB, gtRevision uint64, expectedWritePos uint64) (WatchableList, error) {
+	// multiple subs will compete to fill the cache... only the first in will do the work
+	cache := sub.cache
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+	if expectedWritePos == cache.writePos {
+		// fetch the data, and write it to the ring buffer...
+		w, err := sub.fetchFn(db, gtRevision)
+		if err != nil {
+			return w, err
+		}
+		ringSize := uint64(len(cache.ringBuffer))
+
+		fetchLength := w.Len()
+		if fetchLength > int(ringSize) {
+			// we fetched more than what fits in the ring, NOT ideal.
+			fetchLength = int(ringSize)
+		}
```

Comment:
Entries from the start of the fetch would get overwritten by entries at the tail fetch since the ring buffer is smaller.    Then the worker would try to get the next entry from the ring buffer and see that the next entry is not the next one he is expecting and then we would stop using the ring buffer to fetch and go back to polling.